### ISSUE TITLE
Postgrest returns empty status when new-lines are in JSON payload

### ISSUE THREAD

**Issue by ekosz** [begriffs/postgrest#456]
When I try to create a record with a payload with multiple new lines like this:

```
POST /updates

{
  "comment": "Foobar\n\nbarbar" 
}
```

Postgrest then returns an empty response. As in no headers or anything. On NodeJS this acts as a `socket hang up` error. Using Postman is shows up as 

```
This seems to be like an error connecting to http://xxx/updates. The response status was 0.
Check out the W3C XMLHttpRequest Level 2 spec for more details about when this happens.
```

Strangely this only happens with my Postgrest running on Heroku but not the version I have running locally. My local version is `3.0.3`, while it seems the version on Heroku is `3.0.0`. But looking at the Changelog I don't see anything that would address an issue like this one.

Do you know if this is a fixed issue already? If so, is there advice on how to upgrade Postgrest on Heroku while still using the [postgrest-heroku](https://github.com/begriffs/postgrest-heroku) buildpack?

If not, what can I do to help debug this issue for you all?

**Comment by begriffs** [comment]
That's a weird bug. I did change and standardize the way input for POST requests is handled (to make it internally uniform between json and csv) but I did that in v0.3.0.0. Could you try either deploying a new Heroku instance using 0.3.0.3 or upgrading your existing instance?

To upgrade your existing Heroku instance to a new postgrest version:

1. Get the [Heroku Toolbelt](https://toolbelt.heroku.com/)
2. Clone your project, change settings, and push for new deployment.

    ```bash
    # Clone your app locally
    heroku git:clone -a [your app name]
    cd [your app name
    # Set config vars to force an upgrade
    heroku config:set POSTGREST_VER=0.3.0.3
    heroku config:set CLEAR_CACHE=1
    # Add a fake commit and push for deployment
    git commit --allow-empty -m redeploy
    git push heroku master
    ```

**Comment by ekosz** [comment]
I upgraded our staging server. No dice. Still get the same error. Anything else I can try? I've also noticed the amount of logging Postgrest has is pretty non-existant. Is there some debug mode I can try turning on?

**Comment by begriffs** [comment]
Can you confirm that the server is running 0.3.0.3? There should be a `Server` response header with the version in it.

Also try passing the header `Prefer: return=representation` with your post request. That will tell postgrest to echo the created object back to you.

You can run `heroku logs -t` to see logs of the requests coming in. That's as detailed as it gets though. I sometimes look at the database logs to get more idea of what's happening.

**Comment by ekosz** [comment]
```
Server → postgrest/0.3.0.3
```

Adding the `Prefer` header doesn't change the response. In fact looking at the heroku logs I have:

```
2016-01-14T00:38:21.201640+00:00 app[web.1]: 10.139.101.59 - - [14/Jan/2016:00:38:21 +0000] "POST /updates HTTP/1.1" 201 - "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
```

But no corresponding `router` log line. Which I assume means that the Heroku routing layer isn't seeing a response either.

**Comment by begriffs** [comment]
If you run a server locally and point it at the same database does your local server have the same problem?

**Comment by ekosz** [comment]
Oh, good idea. I tried that out and it worked connecting to the staging DB from my local. The only differences I can think of now are:

1. A difference in the binary. OSX vs whatever Heroku runs. 
2. Somehow the Heroku routing layer is doing something to the request. Maybe they have an internal WAF I'm hitting? But that seems unlikely.

**Comment by begriffs** [comment]
Heroku is running the Ubuntu 64 bit binary. I built the various binaries from the same source commit so it's odd.

What happens if you try posting json that you know has a problem, such as a primary key duplication. Then see if you get an error. Try it with and without the newlines in the payload. That may show if your request is not getting through or if the response is being blocked.

**Comment by ekosz** [comment]
Yep it's getting to Postgrest for sure. When I change the payload to:

```json
{
  "wrong": "something",
  "comment": "Foobar\n\nbarbar" 
}
```

I get:

```json
{
    "hint": null,
    "details": null,
    "code": "42703",
    "message": "column \"wrong\" of relation \"updates\" does not exist"
} 
```

Also, all of these requests ARE inserting rows in the DB. It's just the response that is off. I'm starting to think it's Heroku's processing of the Postgrest response which is off. I just don't know why yet.

**Comment by ekosz** [comment]
This is the response I got back from Heroku:

> Hi Eric,

> We did some investigating and it looks like the problem here is that your application is returning an invalid response which is causing our router process to crash. Here's the raw contents of the response:

> ```
> transfer-encoding: chunked
> Date: Fri, 15 Jan 2016 01:47:14 GMT
> Server: postgrest/0.3.0.3
> Content-Type: application/json
> Location: /updates?id=is.null&status=is.null&comment=eq.Foobar   <---- !!! issue is here !!!
>  barbar&increment=is.null&company_id=is.null&goal_id=is.null&creator_id=is.null&created_at=is.null&updated_at=is.null

> 009E
> {"id":null,"status":null,"comment":"Foobar\n\nbarbar","increment":null,"company_id":null,"goal_id":null,"creator_id":null,"created_at":null,"updated_at":null}
> 0
> ```

> The location header contains an unencoded newline, which isn't valid. If you can get that fixed, the issue should stop happening. Let me know if you have any questions.

So it looks like whatever software / hardware they're using it really doesn't like invalid headers. @begriffs do you know if Postgrest is already url encoding the `location` header?

**Comment by begriffs** [comment]
Yep it's a postgrest problem. The server adds the Location header but looks like it needs to properly urlencode it. Thanks for tracking this down, I'll fix it.

As a workaround you can send the request header `Prefer: return=minimal` which will suppress the Location response header.

Related question: does your `updates` table lack a primary key? The location header we're seeing includes filters for all those columns because it can't find a single primary key to use to reference the created resource.

**Comment by ekosz** [comment]
@begriffs The updates table is a view. Is there a way to mark a column as being the "primary" column in a view?

**Comment by begriffs** [comment]
PostgREST can infer foreign key relationships through views to their underlying tables but I don't think it can do that for primary keys. ( @ruslantalpa do you think this kind of detection is theoretically feasible? )

**Comment by calebmer** [comment]
PostgREST does infer primary keys and foreign keys. But only for auto-updatable views.

**Comment by begriffs** [comment]
I've got a failing test locally reproducing your problem. I'm looking into how to fix it. There's a complication in that the entire location link is being [generated in sql](https://github.com/begriffs/postgrest/blob/860e4370782268d23253bd1b5d9912115bf1ceab/src/PostgREST/QueryBuilder.hs#L358-L369).

**Comment by calebmer** [comment]
This should be an easy fix as PostgreSQL comes with a [url_encode](http://stackoverflow.com/questions/10318014/javascript-encodeuri-like-function-in-postgresql#10372828) function.

**Comment by begriffs** [comment]
I made this modification of `locationF` which will return the location header query params as json objects that we can properly quote in postgrest.

```haskell
locationF :: [Text] -> SqlFragment
locationF pKeys =
    "(" <>
    " WITH s AS (SELECT row_to_json(ss) as r from " <> sourceCTEName <> " as ss  limit 1)" <>
    " SELECT array_to_json(array_agg(location_cols))" <>
    " FROM s, json_each_text(s.r) AS location_cols" <>
    (
      if null pKeys
      then ""
      else " WHERE location_cols.key IN ('" <> intercalate "','" pKeys <> "')"
    ) <>
    ")"
```

However I think it will be easier to parse after finishing and merging my hasql-19 branch. @ekosz, would you be OK using the Prefer header workaround while I get that other branch finished?

**Comment by begriffs** [comment]
Oh @ekosz I was also meaning to ask you, what kind of extra logging would you like the app to provide?

**Comment by ekosz** [comment]
@begriffs Things that I would find useful in a debug mode:

* "Connected to postgres DB" when it first starts up just to know that it's working.
* Simple SQL timings. "SELECT FROM 1.updates... 3ms". "INSERT INTO 1.auths.... 20ms". That way I can compare what Postgres is say, Postgrest, and other services down the line and see where network latency is affecting things.
* Lifecycle events. "Parsing request". "Performing query". "Transforming results". "Sending response"

That's what I can think of off the top of my head.

**Comment by ekosz** [comment]
@begriffs Do you know the timeline for your branch? Unfortunately I can't use the `Prefer` header as there are some uuid columns that are generated in Postgres that I need access to.

Thanks!

**Comment by begriffs** [comment]
The branch is so tantalizingly close, but there remains a bug with the test suite and how it uses database connections. For more info see #472. It's driving me nuts because I want to move on to fixing other things.

**Comment by ekosz** [comment]
ping @begriffs how goes the new branch? Looking green! Let me know if there is anything I can do to help, I don't know haskell very well, but I'm happy to learn to help out fixing this issue haha.

Cheers.

**Comment by begriffs** [comment]
That other branch has been merged and I can hopefully fix this issue tomorrow.

**Comment by begriffs** [comment]
Sorry I got started on this and then learned that the deadlock problem reported in another issue is more serious than I thought. Rushing to fix that then back to this problem. I realize it's blocking you so I do apologize.

**Comment by ekosz** [comment]
Hey @begriffs don't worry about me. I switched off of Heroku for this and other issues I was having (latency between services mostly).

**Comment by thatsmydoing** [comment]
I also have this problem with version 0.3.1.1. Inserts without return=minimal gives me 500s with a body of "Something went wrong". Running the centos build on CentOS gives me this problem. Running the centos build on Archlinux is fine, though.

On a whim, I then tried using the ubuntu build on CentOS and it's working fine. I have no idea why.

**Comment by begriffs** [comment]
@thatsmydoing are you connecting to the same database or different databases for these tests on various platforms?

**Comment by thatsmydoing** [comment]
Yes, they were all against the same database. Funnily enough, I tried the centos build again, and it works now. I really don't know what's happening.

For what it's worth, these were my logs from when it still happened
```
Jun 06 13:05:41 hostname postgrest[436]: 127.0.0.1 - - [06/Jun/2016:13:05:41 +0800] "DELETE /announcements HTTP/1.0" 204 - "http://hostname/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.63 Safari/537.36"
Jun 06 13:05:41 hostname postgrest[436]: 127.0.0.1 - - [06/Jun/2016:13:05:41 +0800] "GET /announcements HTTP/1.0" 200 - "http://hostname/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.63 Safari/537.36"
Jun 06 13:05:46 hostname postgrest[436]: ConnectionError (Just "could not connect to server: Connection refused\n\tIs the server running on host \"127.0.0.1\" and accepting\n\tTCP/IP connections on port 5432?\n")
Jun 06 13:06:06 hostname postgrest[436]: ConnectionError (Just "could not connect to server: Connection refused\n\tIs the server running on host \"127.0.0.1\" and accepting\n\tTCP/IP connections on port 5432?\n")
```
The error only happens when I POST something, but not on GET or DELETE. I also didn't have any trouble inserting via command line while postgrest was running, so the database definitely wasn't down.

**Comment by egeste** [comment]
+1 - I'm seeing this behavior as well

**Comment by ruslantalpa** [comment]
Use the latest release which fixes headers problem

**Comment by begriffs** [comment]
This issue has become very broad. Quite likely much of it is solved in the 0.4 release. I'm closing it, but feel free to open new issues for specific items.
